<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working Locally using anaconda, please uncomment the following code and execute it.
Use the version as per your python version.


In [1]:
!pip install yfinance
!pip install bs4
!pip install nbformat
!pip install --upgrade plotly

In [ ]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.io as pio
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import re

In [3]:
import plotly.io as pio
pio.renderers.default = "iframe"

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [4]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. **You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.**


In [5]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()
    from IPython.display import display, HTML
    fig_html = fig.to_html()
    display(HTML(fig_html))

Use the make_graph function that we’ve already defined. You’ll need to invoke it in questions 5 and 6 to display the graphs and create the dashboard. 
> **Note: You don’t need to redefine the function for plotting graphs anywhere else in this notebook; just use the existing function.**


## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [6]:
tsla = yf.Ticker("TSLA")
tsla


yfinance.Ticker object <TSLA>

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [7]:
tesla_data = tsla.history(period="max")  
tesla_data.tail(3)

Open        High         Low       Close  \
Date                                                                       
2025-09-16 00:00:00-04:00  414.50000  423.250000  411.429993  421.619995   
2025-09-17 00:00:00-04:00  415.75000  428.309998  409.670013  425.859985   
2025-09-18 00:00:00-04:00  428.86499  431.549988  421.950012  427.318390   

                              Volume  Dividends  Stock Splits  
Date                                                           
2025-09-16 00:00:00-04:00  104285700        0.0           0.0  
2025-09-17 00:00:00-04:00  105060600        0.0           0.0  
2025-09-18 00:00:00-04:00   19832026        0.0           0.0

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [8]:

tesla_data.reset_index(inplace=True)  

tesla_data["Date"] = pd.to_datetime(tesla_data["Date"])

tesla_data.head()


Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [21]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

resp = requests.get(url, timeout=30)
resp.raise_for_status()           
html_data = resp.text            
len(html_data), html_data[:200]  

(64779,
 '\n<!DOCTYPE html>\n<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->\n<!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]-->\n<!--[if IE 8]>         <html class')

Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [22]:
soup = BeautifulSoup(html_data, "html.parser")
soup.title.get_text(strip=True)

'Tesla Revenue 2010-2022 | TSLA | MacroTrends'

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Step-by-step instructions</summary>

```

Here are the step-by-step instructions:

1. Create an Empty DataFrame
2. Find the Relevant Table
3. Check for the Tesla Quarterly Revenue Table
4. Iterate Through Rows in the Table Body
5. Extract Data from Columns
6. Append Data to the DataFrame

```
</details>


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


In [23]:
%pip install lxml html5lib

all_tables = pd.read_html(html_data)   

cand = [t for t in all_tables if t.shape[1] >= 2 and any("Revenue" in str(c) for c in t.columns)]
if not cand:
    raise RuntimeError("Keine passende Tabelle gefunden.")
tmp = cand[0].copy()


tmp.columns = ["Date", "Revenue", *[f"col{i}" for i in range(tmp.shape[1]-2)]]
tesla_revenue = tmp[["Date", "Revenue"]].copy()

tesla_revenue.head()

Note: you may need to restart the kernel to use updated packages.


Date  Revenue
0  2021  $53,823
1  2020  $31,536
2  2019  $24,578
3  2018  $21,461
4  2017  $11,759

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [24]:
tesla_revenue["Revenue"] = tesla_revenue["Revenue"].str.replace(r",|\$", "", regex=True)


Execute the following lines to remove an null or empty strings in the Revenue column.


In [25]:
tesla_revenue.dropna(subset=["Date", "Revenue"], inplace=True)
tesla_revenue = tesla_revenue[tesla_revenue["Revenue"] != ""]


Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [26]:

year_str = tesla_revenue["Date"].astype(str).str.extract(r"(\d{4})")[0]

tesla_revenue["Date"] = pd.to_datetime(year_str, format="%Y")

tesla_revenue["Revenue"] = pd.to_numeric(tesla_revenue["Revenue"], errors="coerce")
tesla_revenue = tesla_revenue.dropna(subset=["Date","Revenue"]).sort_values("Date").reset_index(drop=True)

tesla_revenue.tail()


Date  Revenue
8  2017-01-01    11759
9  2018-01-01    21461
10 2019-01-01    24578
11 2020-01-01    31536
12 2021-01-01    53823

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [27]:
gme = yf.Ticker("GME")
gme

yfinance.Ticker object <GME>

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to ` "max" ` so we get information for the maximum amount of time.


In [29]:
gme_data = gme.history(period="max")
gme_data.tail(3) 

Open       High        Low  Close   Volume  \
Date                                                                         
2025-09-16 00:00:00-04:00  25.690001  26.420000  25.610001  26.34  9830000   
2025-09-17 00:00:00-04:00  26.350000  26.469999  25.580000  26.10  7843600   
2025-09-18 00:00:00-04:00  26.209999  26.416500  26.070000  26.25  1671756   

                           Dividends  Stock Splits  
Date                                                
2025-09-16 00:00:00-04:00        0.0           0.0  
2025-09-17 00:00:00-04:00        0.0           0.0  
2025-09-18 00:00:00-04:00        0.0           0.0

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [32]:
gme_data.reset_index(inplace=True)
gme_data["Date"] = pd.to_datetime(gme_data["Date"])
gme_data.head()


level_0  index                      Date      Open      High       Low  \
0        0      0 2002-02-13 00:00:00-05:00  1.620129  1.693350  1.603296   
1        1      1 2002-02-14 00:00:00-05:00  1.712707  1.716073  1.670626   
2        2      2 2002-02-15 00:00:00-05:00  1.683250  1.687458  1.658001   
3        3      3 2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578047   
4        4      4 2002-02-20 00:00:00-05:00  1.615921  1.662210  1.603296   

      Close    Volume  Dividends  Stock Splits  
0  1.691667  76216000        0.0           0.0  
1  1.683250  11021600        0.0           0.0  
2  1.674834   8389600        0.0           0.0  
3  1.607504   7410400        0.0           0.0  
4  1.662210   6892800        0.0           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data_2`.


In [36]:
url_gme = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

resp = requests.get(url_gme, timeout=30)
resp.raise_for_status()
html_data_2 = resp.text   
len(html_data_2), html_data_2[:200]  

(88703,
 '<!DOCTYPE html>\n<!-- saved from url=(0105)https://web.archive.org/web/20200814131437/https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue -->\n<html class=" js flexbox canvas canvastext webgl')

Parse the html data using `beautiful_soup` using parser i.e `html5lib` or `html.parser`.


In [37]:
soup2 = BeautifulSoup(html_data_2, "html.parser")
soup2.title.get_text(strip=True)  

'GameStop Revenue 2006-2020 | GME | MacroTrends'

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column.


> **Note: Use the method similar to what you did in question 2.**  


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [40]:
tables = pd.read_html(html_data_2, match="GameStop")
if not tables:
    
    tables = pd.read_html(html_data_2)
    tables = [t for t in tables if t.shape[1] >= 2 and any("Revenue" in str(c) for c in t.columns)]
    if not tables:
        raise RuntimeError("Keine passende Tabelle gefunden.")
tmp = tables[0].copy()
tmp.columns = ["Date","Revenue", *[f"col{i}" for i in range(tmp.shape[1]-2)]]
gme_revenue = tmp[["Date","Revenue"]].copy()
gme_revenue.head()


Date Revenue
0  2020  $6,466
1  2019  $8,285
2  2018  $8,547
3  2017  $7,965
4  2016  $9,364

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [41]:

gme_revenue["Revenue"] = gme_revenue["Revenue"].astype(str).str.replace(r",|\$", "", regex=True)


gme_revenue = gme_revenue[gme_revenue["Revenue"].ne("")].dropna(subset=["Date","Revenue"])


year_str = gme_revenue["Date"].astype(str).str.extract(r"(\d{4})")[0]
gme_revenue["Date"] = pd.to_datetime(year_str, format="%Y")


gme_revenue["Revenue"] = pd.to_numeric(gme_revenue["Revenue"], errors="coerce")


gme_revenue = gme_revenue.dropna(subset=["Date","Revenue"]).sort_values("Date").reset_index(drop=True)

gme_revenue.tail()


Date  Revenue
11 2016-01-01     9364
12 2017-01-01     7965
13 2018-01-01     8547
14 2019-01-01     8285
15 2020-01-01     6466

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


In [ ]:

pio.renderers.default = "notebook_connected"

def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(
        rows=2, cols=1, shared_xaxes=True,
        subplot_titles=("Historical Share Price", "Historical Revenue"),
        vertical_spacing=0.3
    )
   
    stock_data_specific = stock_data[stock_data["Date"] <= "2021-06-14"].copy()
    revenue_data_specific = revenue_data[revenue_data["Date"] <= "2021-04-30"].copy()

    fig.add_trace(
        go.Scatter(
            x=pd.to_datetime(stock_data_specific["Date"]),
            y=stock_data_specific["Close"].astype(float),
            name="Share Price"
        ),
        row=1, col=1
    )
    fig.add_trace(
        go.Scatter(
            x=pd.to_datetime(revenue_data_specific["Date"]),
            y=revenue_data_specific["Revenue"].astype(float),
            name="Revenue"
        ),
        row=2, col=1
    )

    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)

    fig.update_layout(
        showlegend=False,
        height=900,
        title=stock,
        xaxis_rangeslider_visible=True
    )
    fig.show()


tesla_data["Date"] = pd.to_datetime(tesla_data["Date"])
tesla_revenue["Date"] = pd.to_datetime(tesla_revenue["Date"])

make_graph(tesla_data, tesla_revenue, "Tesla")

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


<details><summary>Hint</summary>

```

You just need to invoke the make_graph function with the required parameter to print the graphs.The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
